Lesson 9

ignatbagaev opened the pull request.

ignatbagaev committed:
create Vote model

ignatbagaev committed:
vote for question

Registered user could upvote or downvote somebody's question but could not vote for own question

ignatbagaev committed:
voting for answer

ignatbagaev committed:
models refactoring

Code refactoring in models with using concerns

ignatbagaev committed:
refactor controllers

ignatbagaev committed:
code cleaning with rubocop

ignatbagaev commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,21 @@
+  end
+
+  def rating
+    "rating: #{positive_votes - negative_votes}"
```
не лучше ли добавить поле raiting и вопросу, и ответу, обновляя это поле при голосовании, чем поступать таким образом?

vkurennov replied:
Да, хранить в моделях можно, но это лишь оптимизация, чтобы не считать каждый раз. Но перед тем, как это сделать, нужно реализовать более важную оптимизацию: переложить подчсет рейтинга на базу. Что происходит у тебя сейчас? Ты делаешь 2 запроса  к базе для положительных и отрицательных голосов, при этом, используется функция count, которая перебирает все нужные строки в таблице. А если значение голоса сделать числовым, 1 и -1, то за один запрос, функцией sum можно сразу получить общий рейтинг. Ну а потом уже можно и в модели эти значения "закешировать"

ignatbagaev committed:
DRYed repetitive code in js

vkurennov commented on app/controllers/concerns/authored.rb:
```diff
@@ -0,0 +1,11 @@
+module Authored
+  private
+
+  def check_author(object)
```
Вообще, этот метод можно было вынести в ApplicationController, т.к. он много где используется, чтобы не подключать модули какждый раз.

vkurennov commented on app/controllers/concerns/authored.rb:
```diff
@@ -0,0 +1,11 @@
+    if current_user.author_of?(object)
+      yield if block_given?
+    else
+      controller_name == 'questions' && action_name == 'destroy' ? (redirect_to object) : (render head: 403)
```
А вот это не очень хорошо - делать хардкод каких-то условий для конткретного контроллера. Можно например, сделать необязательный параметр метода и в нем задавать куда делать редирект (путь), а если на задано, то редиректить по-умолчанию на объект.
А еще лучше не выносить редирект в этот метод, здесь должна быть только проверка или ответ 403, а если все прошло, то решение о том, куда будет редирект лучше написать в контроллере, так будет нагляднее гораздо

vkurennov commented on app/controllers/concerns/voted.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def upvote
+    unless current_user.author_of?(@votable)
+      @vote = @votable.votes.new(positive: true, user_id: current_user.id)
```
1. Не надо по id связывать объекты (по id мы проверяем), т.е. просто `user: current_user`
2. А почему бы не сделать значение голоса числовым? 1 и -1, тогда сумму можно будет считать агрегатной функцией sum и переложить это на БД

vkurennov commented on app/controllers/concerns/voted.rb:
```diff
@@ -0,0 +1,44 @@
+
+  def downvote
+    unless current_user.author_of?(@votable)
+      @vote = @votable.votes.new(positive: false, user_id: current_user.id)
```
Те же замечания ,что и к upvote

vkurennov commented:
Сделай оптимизации о которых я написал и также сделай хранение общего рейтинга в вопросе/ответе, это еще одна оптимизация. Только хранение и обновление рейтинга реализуй через коллбеки в моделях, а не в контроллере. Т.е. при создании голоса или его удалении, должен соответствующим образом обновляться результирующий рейтинг объекта.

ignatbagaev committed:
fixes

ignatbagaev committed:
refactoring

Moved voting logic from controller to model and wrote specs to test the behaviour

ignatbagaev committed:
refactoring in votable module

vkurennov commented on app/controllers/application_controller.rb:
```diff
@@ -2,4 +2,12 @@ class ApplicationController < ActionController::Base
+
+  private
+
+  def user_owns?(object)
```
По большому счету, смысла в этом методе нет, ну разве что, только запись короче, а так - это то же самое, что  и current_user.author_of?(object)

vkurennov commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,30 @@
+
+  def vote!(action, user)
+    transaction do
+      action == 'unvote' ? delete_vote(action, user) : create_vote(action, user)
```
я бы разделил метода создания и удаления голоса, ничего страшного, что в контроллере будет чуть больше кода.

vkurennov commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,30 @@
+  private
+
+  def create_vote(action, user)
+    value = action == 'upvote' ? 1 : -1
```
Не лучшая идея привязываться к называниям экшенов, пусть лучше будет метод голосования "за" и метод голосования "против", в них будут подставляться 1 или -1, а контроллер будет вызывать нужный метод.

vkurennov commented:
У тебя слишком сильно связаны контроллер и модель, модель знает о названии экшенов в контроллере, например. Это не очень хорошо, т.к. классы надо стараться проектировать как можно более слабосвязанными и модели вообще ничего не должны знать о контроллерах, ну и это не гибко. Поменяли название экшена и надо менять модель. Так что, исправь замечания

ignatbagaev committed:
final fixes

vkurennov commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,31 @@
+  def upvote!(user)
+    transaction do
+      votes.create(value: 1, user: user)
+      update(rating: (rating + 1))
```
Кстати, есть метод increment для этого

vkurennov commented on app/models/concerns/votable.rb:
```diff
@@ -0,0 +1,31 @@
+  def downvote!(user)
+    transaction do
+      votes.create(value: -1, user: user)
+      update(rating: (rating + (-1)))
```
А не проще писать rating - 1?

ignatbagaev replied:
@vkurennov, мне показалось, не будет лишним явно указать, что значение голоса прибавляется к рейтингу, т.к. вопрос создаётся

vkurennov commented:
Вот тепреь как надо, принято
